MM-16821 - Add a KVCompareAndDelete to the plugin API

g3rv4 opened the pull request:
Adds KVCompareAndDelete so that plugins don't need to do [ugly things](https://github.com/mattermost/mattermost-plugin-nps/blob/521c8d962167e1d94d0b2f7b3568c90b6ee062fb/server/lock.go#L66-L67) to delete values atomically.

It also adds `KVCompareAndDeleteJSON` to the API.

Fixes https://mattermost.atlassian.net/browse/MM-16821

~I'm heading out for the day, but tomorrow morning I'll add the new methods to the API documentation.~ heh, I just see that it's auto generated from the comments.

g3rv4 commented on plugin/api.go:
```diff
@@ -469,6 +469,14 @@ type API interface {
+	// Returns (false, nil) if current value != oldValue or key does not exist when deleting
+	// Returns (true, nil) if current value == oldValue and the key was deleted
+	//
+	// Minimum server version: TBD
```
When is this determined?

lieut-data replied:
We have a tick-tock release cycle, with all changes in `master` currently slated for the upcoming 5.16 release in October.

g3rv4 commented on plugin/helpers.go:
```diff
@@ -22,6 +22,14 @@ type Helpers interface {
+	// Returns (false, nil) if current value != oldValue or the key was already deleted
+	// Returns (true, nil) if current value == oldValue
+	//
+	// Minimum server version: TBD
```
When is this determined?

g3rv4 commented on store/sqlstore/plugin_store.go:
```diff
@@ -111,6 +111,40 @@ func (ps SqlPluginStore) CompareAndSet(kv *model.PluginKeyValue, oldValue []byte
+	}
+
+	query := `DELETE FROM PluginKeyValueStore WHERE PluginId = :PluginId AND PKey = :Key AND PValue = :Old`
+	if oldValue == nil {
```
do we care about this? I'm not sure it's a valid case for a plugin to store `nil`, since `KVGet` would return `nil` when it can't find a value.

As far I understand (which is very little :P) a plugin wouldn't be able to tell the difference between "there's no value stored for key `key1`" and "there's a value `nil` stored for `key1`"

lieut-data replied:
Agreed -- and now that I think about it, this was part of the impetus for making `KVCompareAndSet` delete on receiving a `nil` value. I do think having an explicit `KVCompareAndDelete` might still be a good idea.

g3rv4 replied:
I see. I still think that having a separate function is best (easier to discover for consumers of the API) but I’m more than happy to go the original route if you prefer so.

g3rv4 replied:
Thoughts about https://github.com/mattermost/mattermost-server/pull/11804/commits/d17f7c4d1cff47193dfa98e5e9988f58c4300a9c? With this change, when the caller uses `CompareAndSet` with `nil` as the new value, it's deleted.

lieut-data replied:
👍 for doing it on both routes.

lieut-data commented:
@prapti, heads up that this ticket will likely be most easily verified once the NPS plugin is updated to use this new API method.

g3rv4 commented:
@lieut-data @prapti updating the NPS plugin is... a bit trickier than I anticipated, since it was built pointing to [this commit](https://github.com/mattermost/mattermost-server/commit/dce6cb601f15f27cb35fb37b6863ee4626df6d01) of `mattermost-server`, and at that point the server was using [analytics-go 2.0.1](https://github.com/mattermost/mattermost-server/blob/dce6cb601f15f27cb35fb37b6863ee4626df6d01/go.mod#L64). The server is now using [3.0.1](https://github.com/mattermost/mattermost-server/blob/4734215a114ef7fd7dcf4ccf3f72a7568b170229/go.mod#L67) and the new version doesn't have the method the NPS plugin is using.

This change was done [last month](https://github.com/mattermost/mattermost-server/commit/d844c52f06c8c0a12ca67e3cef3c9e6aae275694), after the latest tag. I'm not sure what's the appropriate way of moving forward... should I change the NPS plugin to use what's currently on `master`, so that I can add a fix to it?

Should I cherry pick my changes from the latest tag? I'm not convinced that's good, since the plugin would be built against a version of the server that doesn't really exist. I could do the latter just to prove this works, but I can't really find a nice solution.

lieut-data commented:
@hmhealey, thoughts on @g3rv4's feedback above? It's unfortunate that `analytics-go` isn't using Go modules, otherwise we could theoretically have both packages compiling at the same time, but I suspect the right step forward is just to update the NPS plugin to compile against the `master` version of mattermost-server.

marianunez commented on store/sqlstore/plugin_store.go:
```diff
@@ -111,6 +116,40 @@ func (ps SqlPluginStore) CompareAndSet(kv *model.PluginKeyValue, oldValue []byte
+	}
+
+	if rowsAffected, err := deleteResult.RowsAffected(); err != nil {
+		// Failed to delete
```
We can remove this comment, I don't believe it adds any value in this case

marianunez commented on store/sqlstore/plugin_store.go:
```diff
@@ -111,6 +116,40 @@ func (ps SqlPluginStore) CompareAndSet(kv *model.PluginKeyValue, oldValue []byte
+		return false, model.NewAppError("SqlPluginStore.CompareAndDelete", "store.sql_plugin_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
+	} else if rowsAffected == 0 {
+		// No rows were affected by the delete, where condition was not satisfied,
+		// return false, but no error.
```
Same observation as above  regarding the comments

marianunez commented on app/plugin_key_value_store.go:
```diff
@@ -68,6 +68,26 @@ func (a *App) CompareAndSetPluginKey(pluginId string, key string, oldValue, newV
+	deleted, err := a.Srv.Store.Plugin().CompareAndDelete(kv, oldValue)
+	if err != nil {
+		mlog.Error("Failed to compare and delete plugin key value", mlog.String("plugin_id", pluginId), mlog.String("key", key), mlog.Err(err))
+		return deleted, err
```
Should we do `return nil, err` here instead?

g3rv4 replied:
it should be `false, err`, right? I totally agree :)

marianunez replied:
Yes, I meant `return false, err` 🤦‍♀️

ali-farooq0 replied:
the benefit of returning `deleted, err` here is then the logic that determines what `deleted` is remains in one place, i.e. inside `a.Srv.Store.Plugin().CompareAndDelete(kv, oldValue)`. If we decide to be explicit about this then maybe we should update `return updated, err` to `return false, err` to be consistent.

[0/5](https://docs.mattermost.com/process/training.html#id8) on this.

lieut-data replied:
👍 	for consistency with the existing `return updated, err` pattern.

g3rv4 replied:
thanks @ali-farooq0, I agree that we should be consistent if we go the `return false, err` logic (which personally I prefer). But you're talking about [L60](https://github.com/mattermost/mattermost-server/pull/11804/files#diff-afd8e004133130ca459a180c0ed896fdR60) (not [L68](https://github.com/mattermost/mattermost-server/pull/11804/files#diff-afd8e004133130ca459a180c0ed896fdL68)), right?

g3rv4 replied:
seeing that there seems to be agreement that `updated, err` is better than `false, err` I went with that approach on https://github.com/mattermost/mattermost-server/pull/11804/commits/9ef6d9a45dc35fdd534deac8fa718314a1eed1ca

g3rv4 commented:
@marianunez comments addressed :) thanks!

ali-farooq0 commented:
Thanks for the contribution @g3rv4! One minor nit https://github.com/mattermost/mattermost-server/pull/11804#discussion_r312172308

g3rv4 commented:
Thanks @ali-farooq0! Addressed your comment on https://github.com/mattermost/mattermost-server/pull/11804/commits/9ef6d9a45dc35fdd534deac8fa718314a1eed1ca
